yew-functional-macro/src/lib.rs

```diff
@@ -0,0 +1,145 @@
+use proc_macro2::Span;
+use quote::{format_ident, quote};
+use syn::parse::{Parse, ParseStream};
+use syn::{parse_macro_input, Block, FnArg, Ident, Item, Type};
+
+struct FunctionalComponent {
+    body: Box<Block>,
+    props_type: Type,
+    props_name: Ident,
+}
+
+impl Parse for FunctionalComponent {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let parsed: Item = input.parse()?;
+
+        match parsed {
+            Item::Fn(func) => {
+                if !func.sig.generics.params.is_empty() {
+                    // TODO maybe find a way to handle those
+                    return Err(syn::Error::new_spanned(
+                        func.sig.generics,
+                        "functional components cannot contain generics",
+                    ));
+                }
+
+                let inputs = &func.sig.inputs;
+
+                if inputs.len() > 1 {
+                    return Err(syn::Error::new_spanned(
+                        inputs,
+                        "functional components must take only parameter of props",
+                    ));
+                }
+
+                let (props_type, props_name) = if let Some(arg) = inputs.into_iter().next() {
+                    match arg {
+                        FnArg::Typed(arg) => {
+                            let ident = match &*arg.pat {
+                                syn::Pat::Ident(ident) => &ident.ident,
+                                pat => {
+                                    return Err(syn::Error::new_spanned(
+                                        pat,
+                                        "Cannot obtain ident. This should never happen",
```

Comment:
That can't happen in this case though. 

```rust
pub fn name(ref props: Props) -> Html;
```
would fail because `Props` must be passed as `&Props`. This is fine as is, I guess?

Reply:
No it's not, because `ref props: &Props` is valid.